langchain/retrievers/self_query/base.py

```diff
@@ -76,6 +76,23 @@ def get_relevant_documents(self, query: str) -> List[Document]:
         )
         if self.verbose:
             print(structured_query)
+
```

Comment:
@dev2049 we probably want to fix this upstream rather than here

Reply:
So, the actual generation of the structured query happens at `langchain/chains/query_constructor/base.py`:
```python
def load_query_constructor_chain(
    llm: BaseLanguageModel,
    document_contents: str,
    attribute_info: List[AttributeInfo],
    examples: Optional[List] = None,
    allowed_comparators: Optional[Sequence[Comparator]] = None,
    allowed_operators: Optional[Sequence[Operator]] = None,
    **kwargs: Any,
) -> LLMChain:
    prompt = _get_prompt(
        document_contents, # NOTE: Input goes here...
        attribute_info, # NOTE: Schema for database metadata
        examples=examples,
        allowed_comparators=allowed_comparators, # NOTE: equals, greater than, less than, etc.
        allowed_operators=allowed_operators, # NOTE: and, or, not, potentially extensible list
    )
    print('prompt: ', prompt)
    return LLMChain(llm=llm, prompt=prompt, **kwargs)
```

This output (the LLM-generated structured query) gets passed to
```python
class SelfQueryRetriever(BaseRetriever, BaseModel):
    # [...]
    @classmethod
        def from_llm(
            cls,
            llm: BaseLanguageModel,
            vectorstore: VectorStore,
            document_contents: str,
            metadata_field_info: List[AttributeInfo],
            structured_query_translator: Optional[Visitor] = None,
            chain_kwargs: Optional[Dict] = None,
            enable_limit: bool = False,
            **kwargs: Any,
        ) -> "SelfQueryRetriever":
            if structured_query_translator is None: # NOTE: this translator is vectorstore-specific
                structured_query_translator = _get_builtin_translator(vectorstore.__class__)
            chain_kwargs = chain_kwargs or {}
    
            if "allowed_comparators" not in chain_kwargs:
                chain_kwargs[
                    "allowed_comparators"
                ] = structured_query_translator.allowed_comparators
            if "allowed_operators" not in chain_kwargs:
                chain_kwargs[
                    "allowed_operators"
                ] = structured_query_translator.allowed_operators
            llm_chain = load_query_constructor_chain( # NOTE: Here is where we call load_query_constructor_chain
                llm,
                document_contents,
                metadata_field_info,
                enable_limit=enable_limit,
                **chain_kwargs,
            )
            return cls(
                llm_chain=llm_chain, # NOTE: here is where we use the predicted structured query
                vectorstore=vectorstore,
                structured_query_translator=structured_query_translator, # NOTE: the predicted structured query gets parsed in a vectorstore-specific manner
                **kwargs,
            )
```

My first attempt to fix this upstream, accordingly, was to adjust the prompt that is actually called by `load_query_constructor_chain`, to try to direct the LLM to respect compound operators such as 'AND' and 'OR'. See [below](https://github.com/hwchase17/langchain/pull/5570#discussion_r1213585025). 

Since this did not achieve the desired effect, I'm still left wondering if there actually is a better place to fix this issue upstream. Perhaps some vectorstores can gracefully handle compound operators with only one argument (e.g., 'OR [sole condition]'), but that is unexpected behaviour, in my opinion. 

Another spot that might make sense is `visit_structured_query`, in this function:

```python
    def get_relevant_documents(self, query: str) -> List[Document]:
        """Get documents relevant for a query.

        Args:
            query: string to find relevant documents for

        Returns:
            List of relevant documents
        """
        inputs = self.llm_chain.prep_inputs({"query": query})
        structured_query = cast(
            StructuredQuery, self.llm_chain.predict_and_parse(callbacks=None, **inputs)
        )
        if self.verbose:
            print(structured_query)

        new_query, new_kwargs = self.structured_query_translator.visit_structured_query(
            structured_query
        )
        if structured_query.limit is not None:
            new_kwargs["k"] = structured_query.limit

        search_kwargs = {**self.search_kwargs, **new_kwargs}
        docs = self.vectorstore.search(new_query, self.search_type, **search_kwargs)
        return docs
```

But that function is abstract, and its implementation is defined relative to the different vectorstores. I would hope to not only fix this problem for Chroma, for example, but all types of vectorstores, so this seems like the wrong place to fix it, as well.

The output parser, in turn, unpacks the JSON response from the LLM output. This might be a logical place to process the actual content of the response, but I suspect it makes more sense to let that module handle extracting the query and filter JSON only.

Anyways, what I'm left with at this point is actually my original implementation, in which `SelfQueryRetriever`'s `get_relevant_documents` method recognizes that it cannot use compound operators with one argument. 

I think this fits here given the current implementation, because `get_relevant_documents` actually manipulates the arguments already:

```python
new_query, new_kwargs = self.structured_query_translator.visit_structured_query(
    structured_query 
) # NOTE: can't handle the compound operators here, because visit_structured_query is abstract
if structured_query.limit is not None:
    new_kwargs["k"] = structured_query.limit # NOTE: direct manipulation of arguments already taking place

search_kwargs = {**self.search_kwargs, **new_kwargs}
docs = self.vectorstore.search(new_query, self.search_type, **search_kwargs)
return docs
```

Anyways, that's the result of my digging: back where I started. I think the alternative is to update the StructureQuery/FilterDirective types to recognize two types of operators, compound (AND/OR) and unary (NOT). That would mean updating all of the prompt templates, visitor functions, and structured query parsers for all of the implementations, since the functions are abstract. Happy to look elsewhere though if you have other suggestions!

If not, it would be nice to just work with the current PR for now so I don't have to worry about manually catching the problem in my implementation. We could always open an issue for a more thorough follow up.

Reply:
One other idea I have is to try modifying the prompt template to include negative examples as well, which would include something like `and(eq({'artist': 'Taylor Swift'}))` with an explanation of what's wrong. However, my concern is that if you switch LLMs another model might not catch on easily, even if I could get it working with OpenAI. 

Manually setting the 'k' value and dropping unnecessary wrappers together at filter implementation seems like a more resilient solution.